src/database/size.ts

```diff
@@ -0,0 +1,115 @@
+
+/*
+ * Top-level GET requests for large subtrees will likely
+ * fail, so there are some probable gains to be had from
+ * performing shallow gets first. This constant controls
+ * how deep to search in the tree before attempting large
+ * GETs.
+ */
+const SKIP_DEPTH = 1;
+const TIMEOUT = 50;
+const TIMEOUT_STATUS_CODE = 400;
+
+/*
+ * Control how many outstanding full GETs and shallow
+ * GETs are allowed at a time. Wide trees will have
+ * substantially more "size" operations, whereas deep
+ * trees will have substantially more outstanding list
+ * operations.
+ */
+const SIZE_STACK_CONCURRENCY = 2000;
+const LIST_STACK_CONCURRENCY = 2000;
+
+export class DatabaseSize {
+  path: string;
+  skipDepth: number;
+  sizeEstimate: number;
+
+  listRemote: ListRemote;
+  sizeRemote: SizeRemote;
+
+  private listStack: Stack<() => Promise<string[]>, string[]>;
+  private sizeStack: Stack<() => Promise<SizeResult>, SizeResult>;
+
+  constructor(instance: string, path: string, depth?: number) {
+    this.path = path;
+    this.skipDepth = depth || SKIP_DEPTH;
+    this.sizeEstimate = 0;
+    this.sizeRemote = new RTDBSizeRemote(instance);
+    this.listRemote = new RTDBListRemote(instance);
+    this.listStack = new Stack({
+      name: "list stack",
+      concurrency: LIST_STACK_CONCURRENCY,
+      retries: 3,
+    });
+    this.sizeStack = new Stack({
+      name: "size stack",
+      concurrency: SIZE_STACK_CONCURRENCY,
+      retries: 3,
+    });
+  }
+
+  async execute(): Promise<number> {
+    return this.getSubtreeSize(this.path, 0);
+  }
+
+  private async getSubtreeSize(path: string, depth: number): Promise<number> {
+    let sizeEstimate = 0;
+
+    /*
+     * If we are sufficiently deep in the JSON subtree, try sizing the
+     * entire subtree in a single GET.
+     */
+    if (depth >= this.skipDepth) {
+      const result: SizeResult = await this.sizeStack.run(() =>
+        this.sizeRemote.sizeNode(path, TIMEOUT)
+      );
+      if (result !== null && result.success) {
+        sizeEstimate = result.bytes;
+        return sizeEstimate;
+      }
+      /*
+       * Ignore timeout errors so we can try sizing children of this node.
+       */
+      if (!result.success && result.err.status !== TIMEOUT_STATUS_CODE) {
+        logger.debug(
+          `Unexpected error: '${result.err.message}' when sizing node ${path}. Ignoring.`
+        );
+        throw result.err;
+      }
+    }
+
+    let subPaths: string[] = [];
+    let offset: string;
+
+    do {
+      subPaths = await this.listStack.run(() =>
+        this.listRemote.listPath(path, LIST_BATCH_SIZE, offset, TIMEOUT)
+      );
+      if (subPaths.length === 0) {
+        break;
+      }
+      offset = subPaths[subPaths.length - 1];
+      const sizes: number[] = await Promise.all(
+        subPaths.map((subPath) => this.getSubtreeSize(pathLib.join(path, subPath), depth + 1))
+      );
+      sizeEstimate += sizes.reduce((a, b) => a + b);
+    } while (subPaths.length > 0);
```

Comment:
Is this line reachable?

if `subPaths.length == 0`, it will exit the loop in line 104 already.

Reply:
It would be false if the node we're running a shallow get on has more children than the batch size. The purpose of this loop is to paginate through the shallow get results. The short circuit on line 104 can happen in the final iteration if the number of children is an exact multiple of the batch size.

Reply:
If there are more children left, this should be `true` to do another batch.
The sizing task always ends with a shallow list that returns empty list [line 104] (never at line 111)

So, I prefer `while(true) { ... }` with fewer `break`.